### PR TITLE
fix: shared yAdjustments to Ruler and Dots

### PR DESCRIPTION
**Pull request by noahonyejese**
**This PR**
- This PR resolves the misalignment of the Ruler and HoverDots that indicate the values
- This PR resolves # 110 of the private visualize 

---

**Description**
@bprusinowski I'm not particularly happy w/ this solution, because it's so edge case focused. However let me know what you think the primary reason for this issue is the `app/charts/combo/combo-line-container.tsx`

---

### How to test
1. Go to this [link](https://github.com/visualize-admin/visualization-tool-private/issues/110) 
2. See how the HoverDots and Ruler are misaligned ❌
3. Go to this [link](https://vercel.live/open-feedback/visualization-tool-git-fix-dual-line-chart-hover-ixt1.vercel.app?via=pr-comment-visit-preview-link&passThrough=1) 
4. Create a dual Line Chart
5. Publish the Chart 
6. Decrease the Screen size until the axis titles are atleast 2 lines
7. See how the HoverDots and Ruler are no longer misaligned ✅

### CODE REVIEW

**Review by bprusinowski — COMMENTED** [review]
Thanks @noahonyejese 💯 

When looking at the PR, I started to wonder if in order to reduce the "edge-casiness" we couldn't simply increase the chart top margin so that we don't need to add concept of `yAdjustment`?

You could then scope the changes simply to combo charts inside their `useState` hooks, similarly to how you approached this in https://github.com/visualize-admin/visualization-tool/pull/1885. What do you think, did I miss something?

Also, looks like the tooltip box also needs to be offset to be in the middle between the circles – maybe this would be fixed automatically if we switch to changing `bounds.top`?

Screenshot 2024-11-15 at 16 46 29

**Comment by noahonyejese** [comment]
@bprusinowski do you mean removing the prop `yAdjustment` and copying the calculations into each component? lmk if I understood this correctly.

**Comment by bprusinowski** [comment]
@noahonyejese yes, maybe this logic can simply be inside `useChartState` hooks for relevant charts and add to the `bounds` object we already have?

I meant this with not introducing another concept, as we already have a way to control chart margins. I'd try to see if it's possible to move this here (I might also miss something and this doesn't work 😅).

See that you ultimately do `top: margins.top + axisTitleAdjustment,` in the `Ruler` components, so it looks like maybe adding this value to `bounds.margins.top` would also work?

File: app/charts/combo/combo-line-container.tsx [hunk 1]
```diff
@@ -1,31 +1,44 @@
...
+  const overLappingTitles =
+    axisTitleWidth + otherAxisTitleWidth > bounds.chartWidth;
+
+  const yAdjustment = overLappingTitles
+    ? (axisLabelFontSize + TITLE_VPADDING) * 2
+    : 0;
+
+  useMemo(() => {
+    getYAdjustment?.(yAdjustment);
+  }, [yAdjustment, getYAdjustment]);
```
[the remark this answers is not in the document]
Comment (noahonyejese) [reply app/charts/combo/combo-line-container.tsx:41]: @bprusinowski are you sure about `setYAdjustment` isn't that a bit similar to a react state setter don't you think?

File: app/charts/shared/interaction/ruler.tsx [hunk 2]
```diff
@@ -105,7 +126,7 @@ export const RulerContent = (props: RulerContentProps) => {
         style={{
           height: chartHeight,
           left: xAnchor + margins.left,
-          top: margins.top,
+          top: margins.top + axisTitleAdjustment,
```
[the remark this answers is not in the document]
Comment (noahonyejese) [reply app/charts/shared/interaction/ruler.tsx:129]: You mean in the `combo-line-dual-state.tsx` file f.e? That's where the margins seem to be set, however if so that the only place I see margins beeing calculated 

```javascript
  const margins = getMargins({ left, right, bottom });
  ```
  
  However the problem is that we don't want the margin to be everywhere otherwise the whole thing gets moved down the problem is only specific adjustments for dots, ruler and tooltip

[the remark this answers is not in the document]
Comment (bprusinowski) [reply app/charts/shared/interaction/ruler.tsx:129]: Yes, but now essentially everything is moved down anyway by adding a transform to `ChartContainer` – I think you might need to add another property to the `bounds` object, like `yAxisTitleHeight`, that you can access anywhere in the code (and eliminate a need to introduce new property to `ChartContainer`, `useState` with `yAdjustment`, etc) 👍 

I think by having this in a central place, the tooltip anchor might also get fixed :) Otherwise there needs to be some other approach to make sure it's positioned correctly (same with the bolded X axis values) 👀 

Screenshot 2024-11-21 at 08 37 13

[the remark this answers is not in the document]
Comment (noahonyejese) [reply app/charts/shared/interaction/ruler.tsx:129]: Ah okay i see

**Comment by noahonyejese** [comment]
@bprusinowski I think now it's the way you meant it?

**Review by bprusinowski — APPROVED** [review]
Nice, thanks @noahonyejese, I think it's much better now 🎉 

It would be good to add an entry to `CHANGELOG` to make sure we don't forget about this, I think it's a good habit to try to do it for every PR 👍

**Comment by noahonyejese** [comment]
> Nice, thanks @noahonyejese, I think it's much better now 🎉
> 
> It would be good to add an entry to `CHANGELOG` to make sure we don't forget about this, I think it's a good habit to try to do it for every PR 👍

Yes I've just seen all your updates, sorry I was not really familiar with this so far, I will bind it into my PRs from now on out

**Comment by bprusinowski** [comment]
@noahonyejese no problem at all, I think I forgot to mention it earlier 🫣
